File: postgres-indexed-implementation/engine.go
package main

import (
	"errors"
	"fmt"
	"runtime"
	"sync"
	"time"
)

func (engine *Engine) Evaluate(access AccessLog) (Result, error) {
	if access.AccessID == "" || access.SourceIPv4 == "" || access.URLPath == "" || access.AccessTimestampUTC == "" {
		return Result{}, errors.New("access_id, source_ipv4, url_path, and access_timestamp_utc are required")
	}
	tenantID, departmentID, err := scopeFromPath(access.URLPath)
	if err != nil {
		return Result{}, err
	}
	tenant := engine.tenants[tenantID]
	if tenant == nil {
		return Result{}, fmt.Errorf("no policy exists for tenant %q", tenantID)
	}
	policy := tenant.departments[departmentID]
	if policy == nil {
		policy = tenant.policy
	}
	if policy == nil {
		return Result{}, fmt.Errorf("no policy exists for tenant %q", tenantID)
	}
	parsedTime, err := time.Parse(time.RFC3339Nano, access.AccessTimestampUTC)
	if err != nil {
		return Result{}, fmt.Errorf("invalid access_timestamp_utc: %w", err)
	}
	seconds := parsedTime.UTC().Hour()*3600 + parsedTime.UTC().Minute()*60 + parsedTime.UTC().Second()

	var best candidate
	best = policy.regex[0].match(access.SourceIPv4, best)
	if best.valid() && best.priority == policy.minimum {
		return matchedResult(access.AccessID, policy.id, best), nil
	}
	set := tenant.departmentGroups[departmentID]
	if set == nil {
		set = tenant.tenantGroups
	}
	if set != nil {
		for _, key := range set.keysByAddress[access.SourceIPv4] {
			if value, exists := policy.groupRules[key]; exists {
				best = earlier(best, value)
			}
		}
	}
	if best.valid() && best.priority == policy.minimum {
		return matchedResult(access.AccessID, policy.id, best), nil
	}
	best = policy.regex[1].match(access.URLPath, best)
	best = policy.times.match(seconds, best)
	best = policy.regex[2].match(access.Referer, best)
	best = policy.regex[3].match(access.UserAgent, best)
	if best.valid() {
		return matchedResult(access.AccessID, policy.id, best), nil
	}
	return Result{AccessID: access.AccessID, SelectedPolicyID: policy.id, Action: policy.defaultAction}, nil
}

func matchedResult(accessID, policyID string, value candidate) Result {
	return Result{AccessID: accessID, SelectedPolicyID: policyID, MatchedRuleID: value.ruleID, Action: value.action}
}

func (engine *Engine) EvaluateMany(accesses []AccessLog, workers int) ([]Result, error) {
	if workers <= 0 {
		workers = runtime.GOMAXPROCS(0)
	}
	if workers > len(accesses) {
		workers = len(accesses)
	}
	if workers < 2 || len(accesses) < 512 {
		results := make([]Result, len(accesses))
		for index, access := range accesses {
			value, err := engine.Evaluate(access)
			if err != nil {
				return nil, fmt.Errorf("evaluate %s: %w", access.AccessID, err)
			}
			results[index] = value
		}
		return results, nil
	}
	results := make([]Result, len(accesses))
	var wait sync.WaitGroup
	errorsFound := make(chan error, workers)
	chunk := (len(accesses) + workers - 1) / workers
	for worker := 0; worker < workers; worker++ {
		start := worker * chunk
		end := start + chunk
		if end > len(accesses) {
			end = len(accesses)
		}
		if start >= end {
			break
		}
		wait.Add(1)
		go func() {
			defer wait.Done()
			for index := start; index < end; index++ {
				value, err := engine.Evaluate(accesses[index])
				if err != nil {
					select {
					case errorsFound <- fmt.Errorf("evaluate %s: %w", accesses[index].AccessID, err):
					default:
					}
					return
				}
				results[index] = value
			}
		}()
	}
	wait.Wait()
	select {
	case err := <-errorsFound:
		return nil, err
	default:
		return results, nil
	}
}

File: postgres-indexed-java8-implementation/src/main/java/firewall/Engine.java
package firewall;

import java.time.Instant;
import java.time.LocalDateTime;
import java.time.ZoneOffset;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;
import java.util.concurrent.atomic.AtomicReference;

/**
 * Reads the prebuilt policy index and answers First Match evaluation. The
 * engine is immutable once built so concurrent evaluations are safe.
 */
final class Engine {
    private final Map<String, TenantData> tenants;
    final String dataVersion;

    Engine(Map<String, TenantData> tenants, String dataVersion) {
        this.tenants = tenants;
        this.dataVersion = dataVersion;
    }

    /** Builds a copy of this engine that shares the immutable policy graph but reports a different data version. */
    Engine withDataVersion(String newDataVersion) {
        return new Engine(tenants, newDataVersion);
    }

    EvalResult evaluate(AccessLog access) {
        if (access.accessId == null || access.accessId.isEmpty()
                || access.sourceIpv4 == null || access.sourceIpv4.isEmpty()
                || access.urlPath == null || access.urlPath.isEmpty()
                || access.accessTimestampUtc == null || access.accessTimestampUtc.isEmpty()) {
            throw new IllegalArgumentException("access_id, source_ipv4, url_path, and access_timestamp_utc are required");
        }
        String[] scope = scopeFromPath(access.urlPath);
        String tenantId = scope[0];
        String departmentId = scope[1];
        TenantData tenant = tenants.get(tenantId);
        if (tenant == null) {
            throw new IllegalArgumentException("no policy exists for tenant " + tenantId);
        }
        Policy policy = tenant.departments.get(departmentId);
        if (policy == null) {
            policy = tenant.policy;
        }
        if (policy == null) {
            throw new IllegalArgumentException("no policy exists for tenant " + tenantId);
        }
        int seconds = secondsOfDayUtc(access.accessTimestampUtc);

        Candidate best = policy.regex[Policy.SOURCE_IPV4].match(access.sourceIpv4, Candidate.EMPTY);
        if (best.isValid() && best.priority() == policy.minimum) {
            return matched(access.accessId, policy.id, best);
        }
        GroupSet set = tenant.departmentGroups.get(departmentId);
        if (set == null) {
            set = tenant.tenantGroups;
        }
        if (set != null) {
            List<String> keys = set.keysByAddress.get(access.sourceIpv4);
            if (keys != null) {
                for (String key : keys) {
                    Candidate value = policy.groupRules.get(key);
                    if (value != null) {
                        best = Candidate.earlier(best, value);
                    }
                }
            }
        }
        if (best.isValid() && best.priority() == policy.minimum) {
            return matched(access.accessId, policy.id, best);
        }
        best = policy.regex[Policy.URL_PATH].match(access.urlPath, best);
        best = policy.times.match(seconds, best);
        best = policy.regex[Policy.REFERER].match(access.referer == null ? "" : access.referer, best);
        best = policy.regex[Policy.USER_AGENT].match(access.userAgent == null ? "" : access.userAgent, best);
        if (best.isValid()) {
            return matched(access.accessId, policy.id, best);
        }
        return new EvalResult(access.accessId, policy.id, null, policy.defaultAction);
    }

    private static EvalResult matched(String accessId, String policyId, Candidate candidate) {
        return new EvalResult(accessId, policyId, candidate.ruleId(), candidate.action());
    }

    static String[] scopeFromPath(String path) {
        if (path.length() < 4 || path.charAt(0) != '/') {
            throw new IllegalArgumentException("URL path must begin /{tenant_id}/{department_id}/: " + path);
        }
        int first = indexOfByte(path, '/', 1);
        if (first <= 1) {
            throw new IllegalArgumentException("URL path must begin /{tenant_id}/{department_id}/: " + path);
        }
        int second = indexOfByte(path, '/', first + 1);
        if (second <= first + 1) {
            throw new IllegalArgumentException("URL path must begin /{tenant_id}/{department_id}/: " + path);
        }
        return new String[]{path.substring(1, first), path.substring(first + 1, second)};
    }

    private static int indexOfByte(String value, char target, int start) {
        for (int i = start; i < value.length(); i++) {
            if (value.charAt(i) == target) {
                return i;
            }
        }
        return -1;
    }

    private static int secondsOfDayUtc(String timestamp) {
        Instant instant = Instant.parse(timestamp);
        LocalDateTime ldt = LocalDateTime.ofInstant(instant, ZoneOffset.UTC);
        return ldt.toLocalTime().toSecondOfDay();
    }

    List<EvalResult> evaluateMany(List<AccessLog> accesses, int workers) {
        if (workers <= 0) {
            workers = Runtime.getRuntime().availableProcessors();
        }
        if (workers > accesses.size()) {
            workers = accesses.size();
        }
        if (workers < 2 || accesses.size() < 512) {
            List<EvalResult> results = new ArrayList<>(accesses.size());
            for (AccessLog access : accesses) {
                results.add(evaluate(access));
            }
            return results;
        }
        EvalResult[] results = new EvalResult[accesses.size()];
        ExecutorService executor = Executors.newFixedThreadPool(workers);
        try {
            List<Future<?>> futures = new ArrayList<>(workers);
            int chunk = (accesses.size() + workers - 1) / workers;
            AtomicReference<Throwable> firstError = new AtomicReference<>();
            for (int w = 0; w < workers; w++) {
                int start = w * chunk;
                int end = Math.min(start + chunk, accesses.size());
                if (start >= end) {
                    break;
                }
                futures.add(executor.submit(() -> {
                    for (int i = start; i < end; i++) {
                        try {
                            results[i] = evaluate(accesses.get(i));
                        } catch (Throwable t) {
                            firstError.compareAndSet(null, t);
                        }
                    }
                }));
            }
            for (Future<?> f : futures) {
                try {
                    f.get();
                } catch (Exception e) {
                    firstError.compareAndSet(null, e);
                }
            }
            Throwable error = firstError.get();
            if (error != null) {
                throw new RuntimeException(error);
            }
            return Arrays.asList(results);
        } finally {
            executor.shutdownNow();
        }
    }
}
